### PR TITLE
gh-100414: Add SQLite backend to dbm

### PR DESCRIPTION
**Pull request by erlend-aasland**
Initial version authored by Raymond in https://github.com/python/cpython/issues/100414#issuecomment-1368273140.

* Issue: gh-100414

----
📚 Documentation preview 📚: https://cpython-previews--114481.org.readthedocs.build/

### CODE REVIEW

**Comment by erlend-aasland** [comment]
cc. @presidento

**Comment by erlend-aasland** [comment]
I suggest to focus this PR on the basic, required functionality _only_. We can optimise things in follow-up PRs.

File: Doc/library/dbm.rst
Comment (erlend-aasland) [inline Doc/library/dbm.rst]: All of this is duplicated for each dbm submodule. Instead of copying the text yet another time, set up replacements and use these instead. In a follow-up docs PR, we can reuse these for the other modules.

Comment (erlend-aasland) [reply Doc/library/dbm.rst]: Added in 8c265408c51609c6b4a6788cac9cc5fea7a14888.

File: Doc/library/dbm.rst
Comment (erlend-aasland) [inline Doc/library/dbm.rst]:
```suggestion
```

File: Doc/library/dbm.rst
Comment (erlend-aasland) [inline Doc/library/dbm.rst]:
```suggestion
      .. csv-table::
         :header: "Value", "Meaning"

         ``'r'`` (default), |flag_r|
         ``'w'``, |flag_w|
         ``'c'``, |flag_c|
         ``'n'``, |flag_n|
```

**Comment by erlend-aasland** [comment]
I think the basic functionality is ready for an initial round of reviews.

## Tests
There's a basic test suite in `Lib/test_dbm.py`, and I also added some `dbm.sqlite3` specific tests in `Lib/test_dbm_sqlite3.py`. We can expand the latter in follow-up PRs.

## Docs
I also added som basic docs; unless significant information is missing, I suggest we work on the docs in follow-up PRs.

File: Lib/dbm/__init__.py
Comment (gpshead) [inline Lib/dbm/__init__.py]: People using `dbm` are only likely to use a single backend. It is better to delay the import of the specific implementation until it is going to be used.  I'd just always add dbm.sqlite3 to _names above and adjust the loop checking for ImportError within `open()` below if neecessary at all.  An ImportError coming from the attempt to load `dbm.sqlite3` when someone needs or asks for it is sufficient.

Comment (erlend-aasland) [reply Lib/dbm/__init__.py]: Hm, yeah. That's also similar to how the other backends are handled; for example, `dbm.gnu` is always present in `_names`, even though not all systems have the `_gdbm` module.

Comment (erlend-aasland) [reply Lib/dbm/__init__.py]: This reminds me we'd better have some more tests for `.whichdb()`, to make sure SQLite databases are autodetected properly.

Comment (erlend-aasland) [reply Lib/dbm/__init__.py]: Addressed in b655473dbd199006945c2aab8e754684ba7f1fec and 95eba9348c8939cc714881b2ac08648aaf0bfad6.

**Comment by erlend-aasland** [comment]
I added 34d8c7f7e9d385c93563e0fdd1f49b7191234703 to close cursors explicitly for each query. Perhaps it is too cautious. There's plenty of time until the next alpha to tweak the implementation.

**Comment by erlend-aasland** [comment]
@felixxm, are you interested in reviewing this?

File: Lib/dbm/sqlite3.py
Comment (corona10) [inline Lib/dbm/sqlite3.py]: Hmm, since the key is the primary key, IMO, it should guarantee to return just one without limit.

Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: True that

**Review by corona10 — APPROVED** [review]
@erlend-aasland 
For me LGTM.
nit: Can we add the test for verification of unique key?
Something like..
```python
db["key"] = "foo"
db["key"] = "bar"
self.assertEqual(len(db), 1)
self.assertEqual(db["key"], "bar")
```

**Comment by corona10** [comment]
> Yes, let's do that. I wonder if we should change the schema slightly. SQLite works best with [INTEGER PRIMARY KEY](https://www.sqlite.org/autoinc.html).

Hmm, but the key cannot be int type, no? Or Are you going to cast key type into int type by force?

**Comment by erlend-aasland** [comment]
> > Yes, let's do that. I wonder if we should change the schema slightly. SQLite works best with [INTEGER PRIMARY KEY](https://www.sqlite.org/autoinc.html).
> 
> Hmm, but the key is not integer type in some cases. Or Are you going to cast key type into integer type by force?

No, I was thinking of letting SQLite manage the primary key implicitly and just making sure that key is marked unique. However, this optimisation can come in a follow-up PR; we don't have to optimise the database schema in this PR.

File: Lib/dbm/sqlite3.py
Comment (erlend-aasland) [inline Lib/dbm/sqlite3.py]: IMO, we should let SQLite manage the primary key through the implicit `rowid`.

```suggestion
    key TEXT UNIQUE NOT NULL,
    value BLOB NOT NULL
```

Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: cc. @corona10

Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: - https://www.sqlite.org/lang_createtable.html

Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: I'll add some tests for key uniqueness and validity of the generated database.

File: Lib/dbm/__init__.py
Comment (serhiy-storchaka) [inline Lib/dbm/__init__.py]: It makes dbm.sqlite3 the most preferable backend. Is it always better than dbm.gnu and dbm.ndbm?

File: Lib/dbm/sqlite3.py
Comment (serhiy-storchaka) [inline Lib/dbm/sqlite3.py]: It should be documented.

File: Lib/dbm/sqlite3.py
Comment (serhiy-storchaka) [inline Lib/dbm/sqlite3.py]: What if filename contains "%"?

File: Lib/dbm/sqlite3.py
Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]:
```suggestion
LOOKUP_KEY = "SELECT value FROM Dict WHERE key = ?"
```

File: Lib/dbm/sqlite3.py
Comment (serhiy-storchaka) [inline Lib/dbm/sqlite3.py]: It is not efficient and is not atomic. Can we determine whether deletion actually deleted any key by the query result?

File: Lib/dbm/sqlite3.py
Comment (serhiy-storchaka) [inline Lib/dbm/sqlite3.py]: What happen when we close `_Database` in the middle of iteration? What happen when we close it before the first iteration? Are there tests for this?
```py
with closing(iter(db)) as it:
    db.close()
    next(it)  # ???
```
```py
with closing(iter(db)) as it:
    next(it)
    db.close()
    next(it)  # ???
```

File: Lib/dbm/sqlite3.py
Comment (serhiy-storchaka) [inline Lib/dbm/sqlite3.py]: Should there be a formatting operation?

File: Lib/dbm/__init__.py
Comment (erlend-aasland) [reply Lib/dbm/__init__.py]: Yes, this is also one of Raymond's selling points in the OP (as I read him). I definitely think it should be the preferred backend.

File: Lib/dbm/sqlite3.py
Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: Yes, it should; thanks!

File: Lib/dbm/sqlite3.py
Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: Let's find out; I implemented ~~all~~ most of SQLite's recommendations for URI normalisation: https://sqlite.org/uri.html

I see I did not address the backslash case; I'll fix that and add more tests.

File: Lib/dbm/sqlite3.py
Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: Yeah, I think we can implement this better; I'll address it. Thanks.

File: Lib/dbm/sqlite3.py
Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: Let's add those tests.

File: Lib/dbm/sqlite3.py
Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: Yes, of course, this is not `print()` :)

File: Lib/dbm/sqlite3.py
Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: Ah, there is already generic tests for key uniqueness in `test_dbm.py`; we're covered.

File: Lib/dbm/sqlite3.py
Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: > Can we determine whether deletion actually deleted any key by the query result?

We cannot use the RETURNING clause, as that is only available for SQLite 3.35.0 and newer. We _could_ verify it by querying again, but I don't think it is worth it; if DELETE FROM fails, we'll get an operational error, which in turn will show up as an exception for the user.

File: Lib/dbm/sqlite3.py
Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: Perhaps we should not allow it; this is not sqlite3, it is dbm. I suggest we remove this "feature" and treat everything as an ordinary filename.

Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: There is enough time till the next alpha to decide if we want this or not; let's leave it out for now, at least.

Comment (serhiy-storchaka) [reply Lib/dbm/sqlite3.py]: Also, does it work for filenames like "localhost" or "/localhost"?

File: Lib/dbm/__init__.py
Comment (serhiy-storchaka) [reply Lib/dbm/__init__.py]: I think that Raymond said that it will be better than dbm.dumb, not other specialized engines.

Comment (presidento) [reply Lib/dbm/__init__.py]: IMHO Sqlite is cross-platform (and Python has a many users with Windows), although likely not as fast as dbm.gnu or dbm.ndbm, since they are specialized engines especially for this use case.

So what does "better" mean? If better compatibility then Sqlite should be the first. If better performance, then it can be just before dbm.dumb.

For me both are acceptable, these are just my 50 cents.

File: Lib/dbm/sqlite3.py
Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: > Also, does it work for filenames like "localhost" or "/localhost"?

Those are not special names; they will be accepted as ordinary file names.

File: Lib/dbm/__init__.py
Comment (erlend-aasland) [reply Lib/dbm/__init__.py]: > I think that Raymond said that it will be better than dbm.dumb, not other specialized engines.

This is the wording from the OP; I let it speak for itself:

> Right now we support ndbm and gnu.dbm which might or might not be part of a given build. The fallback is the super slow dumbdbm. Not the sqlite3 is part of the standard build, we can do better.

Ping Raymond if you are unsure.

File: Lib/dbm/sqlite3.py
Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: I rewrote this using pathlib and also added substitution for "%".

File: Lib/dbm/sqlite3.py
Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: Implemented by checking `Cursor.rowcount` after the attempted delete operation. Works swell.

File: Lib/dbm/sqlite3.py
Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: Note that the other implementations do not implement `__iter__`. We do, since it is required by the ABC we're using. Anyway, I added guards for misuse.

File: Lib/dbm/sqlite3.py
Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: Windows paths are ... irritating. ~~I'll have to look at those later; for now, I marked the Windows tests as WIP.~~ Windows tests are enabled; crossing fingers 🤞

File: Lib/dbm/sqlite3.py
Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: Let me take that back; `"localhost"` _is_ considered special by SQLite. Let's add more tests.

File: Lib/dbm/sqlite3.py
Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: Pushed 08c38481c10d8cbfcc9ee02f2133a3285eed216b which just uses pathlib's `as_uri()` with some amendment. I guess it's going to break on Windows. Anyway, I'm calling it a day. Thanks for the input so far!

File: Lib/dbm/__init__.py
Comment (corona10) [reply Lib/dbm/__init__.py]: Hmm basicially dbm.gnu and dbm.ndbm is key/value storage product itself but sqlite3 is more rich product than them, I am also not sure that making sqlite3 to become a most preferrable engine is right choice at this moment.

Comment (erlend-aasland) [reply Lib/dbm/__init__.py]: In terms of user experience, dbm.sqlite3 is the clear winner IMO. It is available for everyone; a user can create a database using `dbm.open()` on a Linux machine, copy the generated database to a Windows or macOS machine, open the database using `dbm.open()`, and it just works. Let's ask on Discourse.

Comment (corona10) [reply Lib/dbm/__init__.py]: IMO, even if we will migrate the most preferable engine to sqlite3, CPython 3.13 should be in the experimental stage. So, the most preferable engine should not be SQLite3 at this moment if we consider the experimental stage.
People can argue about different performance issues or behavior issues that can not be expected at this moment.
After CPython 3.13, we can make the decision with more context.

Comment (erlend-aasland) [reply Lib/dbm/__init__.py]: cc. @rhettinger

File: Lib/dbm/__init__.py
Comment (felixxm) [inline Lib/dbm/__init__.py:168]: I know the current comparison works, but can we use `3\000` to be consistent with [SQLite docs](https://www.sqlite.org/fileformat.html#the_database_header):
```suggestion
    if s16 == b"SQLite format 3\000":
```

File: Lib/test/test_dbm.py
Comment (felixxm) [inline Lib/test/test_dbm.py]: I'm not sure it's worth adding prefixes like `Test that`  since all tests test things :upside_down_face: 
```suggestion
        # Databases created by dbm.sqlite3 are detected correctly.
```

**Review by felixxm — COMMENTED** [review]
@erlend-aasland Thanks :+1: I left only two minor suggestions.

File: Lib/test/test_dbm.py
Comment (felixxm) [inline Lib/test/test_dbm.py]:
```suggestion
        # Existing sqlite3 databases are detected correctly.
```

File: Lib/dbm/__init__.py
Comment (erlend-aasland) [reply Lib/dbm/__init__.py:168]: Hm, I would rather do this instead; it feels more "pythonic":

```suggestion
    if s16 == b"SQLite format 3\x00":
```

But it is not a big issue for me; I'm fine either way.

File: Lib/dbm/__init__.py
Comment (serhiy-storchaka) [inline Lib/dbm/__init__.py]:
```suggestion
The returned object is a dbm.sqlite3, dbm.gnu, dbm.ndbm or dbm.dumb database object, dependent on the
```

File: Lib/dbm/__init__.py
Comment (serhiy-storchaka) [reply Lib/dbm/__init__.py]: If "better" means better compatibility then dbm.dumb should be the first. I do not think Raymond used it in this meaning.

File: Lib/dbm/sqlite3.py
Comment (serhiy-storchaka) [inline Lib/dbm/sqlite3.py:33]: Is it needed?

File: Lib/dbm/sqlite3.py
Comment (serhiy-storchaka) [inline Lib/dbm/sqlite3.py]: It can be simply:
```suggestion
            return self._cx.execute(*args, **kwargs)
```
It is even better to wrap it in `closing()`:
```suggestion
            return closing(self._cx.execute(*args, **kwargs))
```

File: Lib/dbm/sqlite3.py
Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py:33]: Yes.

File: Lib/dbm/sqlite3.py
Comment (erlend-aasland) [reply Lib/dbm/sqlite3.py]: Wrapping in closing is a nice trick; I will apply that.

File: Lib/dbm/__init__.py
Comment (erlend-aasland) [reply Lib/dbm/__init__.py]: In order to unlock this PR, let us go for this:

```suggestion
_names = ['dbm.gnu', 'dbm.ndbm', 'dbm.sqlite3', 'dbm.dumb']
```

I'll create a discussion on Discourse, and we can chose to promote it in a follow-up PR.

Comment (erlend-aasland) [reply Lib/dbm/__init__.py]: I created a poll: https://discuss.python.org/t/new-default-preferred-dbm-backend/44228

File: Doc/library/dbm.rst
Comment (erlend-aasland) [inline Doc/library/dbm.rst:181]: List markup needs a newline

```suggestion
   :param str flag:

```

**Comment by erlend-aasland** [comment]
I made a [core dev poll on Discourse](https://discuss.python.org/t/new-default-preferred-dbm-backend/44228/22) regarding if the SQLite backend should be the default one; 16 out of 21 was in favour (Skip changed his mind post voting). This PR does _not_ make the SQLite backend the default. If we want to change the default, we can do it post merging the feature itself. Another possibility, [suggested by @ethanfurman](https://discuss.python.org/t/new-default-preferred-dbm-backend/44228/23):

> Shouldn’t we make it the default now so it can be tested in the alphas? We can undo the default portion if it turns out to be a problem.

@serhiy-storchaka, should we land this?

**Comment by serhiy-storchaka** [comment]
It seems that currently it is not compatible with other dbm implementations. They all are bytes oriented, but accept strings as input (it is a Python 2 legacy). Most of the dbm tests use string input, so I expect that many user code does it too. But the output always expected to be bytes.

Supporting other types, supported by Sqlite looks attractive, but we implement a backend to dbm, so it should be compatible. We can later add an option to alter the behavior, but it should not be default.

**Comment by erlend-aasland** [comment]
> It seems that currently it is not compatible with other dbm implementations. They all are bytes oriented, but accept strings as input (it is a Python 2 legacy). Most of the dbm tests use string input, so I expect that many user code does it too. But the output always expected to be bytes.
> 
> Supporting other types, supported by Sqlite looks attractive, but we implement a backend to dbm, so it should be compatible. We can later add an option to alter the behavior, but it should not be default.

As discussed in DM, I agree with this. I addressed your concern in e782fad38fa032cc2eb3535dd5a9c1f19488353e.

File: Lib/test/test_dbm_sqlite3.py
Comment (serhiy-storchaka) [inline Lib/test/test_dbm_sqlite3.py:215]: Does it work with non-ASCII strings?

**Review by serhiy-storchaka — APPROVED** [review]
LGTM if all operations still accept strings.

**Comment by erlend-aasland** [comment]
With 34930cb9fdc70eaa51548e4140d62c7444c85992, the dbm.sqlite3 backend now aligns with the other backends; for example, the general code snippet in the dbm docs now works as expected with the new backend.

File: Lib/test/test_dbm_sqlite3.py
Comment (erlend-aasland) [reply Lib/test/test_dbm_sqlite3.py:215]: Do you mean such a test?

```diff
diff --git a/Lib/test/test_dbm_sqlite3.py b/Lib/test/test_dbm_sqlite3.py
index 7bc2a03035..3d2dc5fd55 100644
--- a/Lib/test/test_dbm_sqlite3.py
+++ b/Lib/test/test_dbm_sqlite3.py
@@ -214,6 +214,7 @@ class DataTypes(_SQLiteDbmTests):
         (3.14, b"3.14"),
         ("string", b"string"),
         (b"bytes", b"bytes"),
+        ("æøå", b"\xc3\xa6\xc3\xb8\xc3\xa5"),
     )
 
     def setUp(self):
```

**Comment by erlend-aasland** [comment]
Let's land this, @serhiy-storchaka. There is plenty of opportunities to tweak the implementation and to expand the test suite.

Thanks @smontanaro for the initial idea, @rhettinger for the initial implementation, and everyone else involved for ideas, reviews and critique. Thanks to @serhiy-storchaka for the, as always, extremely thorough review of the feature and the implementation.

**Comment by erlend-aasland** [comment]
> ## ⚠️⚠️⚠️ Buildbot failure ⚠️⚠️⚠️
>
> [...]
>
> Failed tests:
> 
> * test_dbm_sqlite3
> 
> [...]
> ModuleNotFoundError: No module named '_sqlite3'
> ```

Proposed fix:

- #115448

**Pull request merged**
